rpcwallet: default include_watchonly to true for watchonly wallets

jb55 opened the pull request:
Right now it's a bit annoying to deal with watchonly wallets, many rpc commands have an `include_watchonly` argument that needs to be explicitly set. 

Wallets created with `createwallet` can have a `disable_private_keys` parameter, for those wallets we already know that they are watchonly, so there's no reason to have to explicitly ask for it for every command. Instead we check this wallet flag when the `include_watchonly` parameter isn't set.

jb55 committed:
rpcwallet: default include_watchonly to true for watchonly wallets

The logic before would only include watchonly addresses if it was
explicitly set in the rpc argument.

This changes the logic like so:

If the include_watchonly argument is missing, check the
WALLET_FLAG_DISABLE_PRIVATE_KEYS flag to determine if we're working
with a watchonly wallet. If so, default include_watchonly to true.

If the include_watchonly argument is explicit set to false, we still
disable them from the listing. Although this would always return
nothing, it might be still useful in situations where you want to
explicitly filter out watchonly addresses regardless of what wallet
you are dealing with.

Signed-off-by: William Casarin <jb55@jb55.com>

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
+ * We default to true for watchonly wallets if include_watchonly isn't
+ * explicitly set
+ */
+static bool IncludeWatchonly(const UniValue& include_watchonly, CWallet* const pwallet)
```
```suggestion
static bool IncludeWatchonly(const UniValue& include_watchonly, const CWallet& wallet)
```

maflcko reviewed (commented):
Would have to update the documentation as well for the "smart" default value.

achow101 commented:
Concept ACK. I think having `include_watchonly` default to true for actual watchonly wallets is super useful.

jb55 commented:
@MarcoFalke I'm a bit new, which docs specifically would need to be updated?

darosior commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
 }
 
+
+/* Used by RPC commands that include an include_watchonly parameter.
```
`/**`-starting comments [are prefered](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments)

darosior commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
 }
 
+
+/* Used by RPC commands that include an include_watchonly parameter.
```
`/**`-starting comments [are prefered](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments)

jb55 replied:
done

MarcoFalke commented:
> @MarcoFalke I'm a bit new, which docs specifically would need to be updated?

```diff
-                    {"include_watchonly", RPCArg::Type::BOOL, /* default */ "false", "Whether to include watch-only addresses in balance calculation and details[]"},
+                    {"include_watchonly", RPCArg::Type::BOOL, /* default */ ">>>>>>>>>>> update here<<<<<<<<", "Whether to include watch-only addresses in balance calculation and details[]"},

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
 }
 
+
+/* Used by RPC commands that include an include_watchonly parameter.
```
nit: for readability perhaps employ a different verb than "include" here (have, take, etc.)

jb55 replied:
done

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
+
+/* Used by RPC commands that include an include_watchonly parameter.
+ * We default to true for watchonly wallets if include_watchonly isn't
+ * explicitly set
```
nit: s/set/set./

jb55 replied:
done

jonatack reviewed (commented):
Concept ACK. Will test when help docs are updated. Seems this change would merit test coverage?

jb55 committed:
rpcwallet: document include_watchonly default for watchonly wallets

Signed-off-by: William Casarin <jb55@jb55.com>

jb55 commented:
@MarcoFalke I wasn't quite sure what to do for documenting the smart default, let me know if that looks ok

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -732,7 +732,7 @@ static UniValue getbalance(const JSONRPCRequest& request)
                     {"dummy", RPCArg::Type::STR, RPCArg::Optional::OMITTED_NAMED_ARG, "Remains for backward compatibility. Must be excluded or set to \"*\"."},
                     {"minconf", RPCArg::Type::NUM, /* default */ "0", "Only include transactions confirmed at least this many times."},
-                    {"include_watchonly", RPCArg::Type::BOOL, /* default */ "false", "Also include balance in watch-only addresses (see 'importaddress')"},
+                    {"include_watchonly", RPCArg::Type::BOOL, /* default */ "false, true for watchonly wallets", "Also include balance in watch-only addresses (see 'importaddress')"},
```
Suggestion for the RPCHelpMan default messages: `"true for watch-only wallets, otherwise false"`

Grepping related cases in the codebase comes up with similar, e.g. src/wallet/rpcdump.cpp:120

As for "watchonly" vs "watch-only", in the descriptions on the same line it is dasherized. Unless other reviewers have a different opinion, it might be more clear to do the same?

jb55 replied:
yeah that message sounds better. I'll go with `watch-only` as well.

jb55 replied:
done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
+ *  We default to true for watchonly wallets if include_watchonly isn't
+ *  explicitly set.
+ */
+static bool IncludeWatchonly(const UniValue& include_watchonly, const CWallet& pwallet)
```
nit, I'd name this `ParseIncludeWatchonly`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1434,9 +1448,9 @@ UniValue listtransactions(const JSONRPCRequest& request)
-        if(request.params[3].get_bool())
-            filter = filter | ISMINE_WATCH_ONLY;
+
+    if (IncludeWatchonly(request.params[3], *pwallet))
```
nit, add `{ }`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1579,9 +1593,8 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
-    if (!request.params[2].isNull() && request.params[2].get_bool()) {
-        filter = filter | ISMINE_WATCH_ONLY;
-    }
+    if (IncludeWatchonly(request.params[2], *pwallet))
```
nit, add `{ }`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1697,9 +1710,9 @@ static UniValue gettransaction(const JSONRPCRequest& request)
-        if(request.params[1].get_bool())
-            filter = filter | ISMINE_WATCH_ONLY;
+
+    if (IncludeWatchonly(request.params[1], *pwallet))
```
nit, add `{ }`.

promag reviewed (commented):
Concept ACK

> Seems this change would merit test coverage?

Agree with @jonatack. Also add a minor release note?

laanwj commented:
Concept ACK, this definitely seems like the only sensible default to have in that case.

instagibbs commented:
concept ACK, needs tests(as others have said)

jb55 commented:
One thing I missed was rpc's with the `includeWatching` option such as `fundrawtransaction` and `walletcreatefundedpsbt`. I think it makes sense to add the default on those as well.

jonasschnelli commented:
Concept ACK

jb55 committed:
doc: add release note for include_watchonly default changes

Signed-off-by: William Casarin <jb55@jb55.com>

jb55 committed:
tests: functional watch-only wallet tests

These test the new watch-only defaults for rpcs with include_watchonly
and includeWatching options.

Signed-off-by: William Casarin <jb55@jb55.com>

jb55 commented:
Latest updates:

- Added release notes
- Added functional tests for the new defaults

This is the first time I've done either of these things, so let me know if I did anything wrong.

practicalswift commented on doc/release-notes-16383.md:
```diff
@@ -0,0 +1,8 @@
+-----------
+
+RPCs which have an `include_watchonly` argument or `includeWatching`
+option now default to `true` for watch-only wallets. affected RPCs
```
Nit: Should be "Affected"? :-)

jb55 replied:
done

practicalswift commented:
Concept ACK

Good idea!

achow101 commented:
Code review ACK 72eaab073bc747425fe551777154b13a6c4c37c9

promag commented on doc/release-notes-16383.md:
```diff
@@ -0,0 +1,8 @@
+
+RPCs which have an `include_watchonly` argument or `includeWatching`
+option now default to `true` for watch-only wallets. Affected RPCs
+are: `getbalance`, `listreceivedbyaddress`, `listreceivedbylabel`,
```
nit, sort.

promag commented on test/functional/wallet_watchonly.py:
```diff
@@ -0,0 +1,107 @@
+        assert_raises_rpc_error(-4, "Insufficient funds", wo_wallet.fundrawtransaction, rawtx, no_wo_options)
+
+
+
```
nit, remove extra line.

promag commented on test/functional/wallet_watchonly.py:
```diff
@@ -0,0 +1,107 @@
+        txid = def_wallet.sendtoaddress(wo_addr, 1)
+        self.nodes[0].generate(1)
+
+        # getbalance
```
nit, drop comment.

promag reviewed (commented):
ACK 72eaab073bc747425fe551777154b13a6c4c37c9, code review only, didn't look closely to the test.

meshcollider commented:
This makes a lot of sense, Concept ACK

Sjors commented:
ACK 72eaab073bc747425fe551777154b13a6c4c37c9

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
     return avoid_reuse;
 }
 
+
```
Nit: Extraneous newline.

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
+
+    // otherwise return whatever include_watchonly was set to
+    return include_watchonly.get_bool();
+}
```
Can do
```C++
return include_watchonly.isNull()
    ? pwallet.IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS)
    : include_watchonly.get_bool();
```

kallewoof commented on src/wallet/rpcwallet.cpp:
```diff
@@ -52,6 +52,23 @@ static inline bool GetAvoidReuseFlag(CWallet * const pwallet, const UniValue& pa
+    return include_watchonly.get_bool();
+}
+
+
```
Nit: Extraneous newline.

kallewoof reviewed (approved):
ACK 72eaab073bc747425fe551777154b13a6c4c37c9

Lots of unnecessary newlines all over, and a few nits, but merge-worthy IMO!

fanquake reviewed (approved):
ACK 72eaab073bc747425fe551777154b13a6c4c37c9 - I've looked over the changes, they make sense to me. Compiled and ran the tests etc.

Thanks to @kallewoof and @ajtowns (he said he sanity checked on his lunch break) for following up my review prod. 

I know there are some nits here, however this has ACKs, and has been sitting for a while, so I'm going to merge it. The nits can be scooped up by someone in either a follow up PR, or a related change.
